File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/get_event_log.go
package workflowexecution

import (
	"context"

	"github.com/rs/zerolog/log"
	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
	"google.golang.org/protobuf/proto"
)

const (
	defaultEventPageSize = 100
	maxEventPageSize     = 500
)

// GetEventLog returns a paginated list of execution events.
//
// Supports cursor-based pagination via after_sequence and optional filtering
// by event type or task name. Events are returned in sequence_number ascending order.
func (c *WorkflowExecutionController) GetEventLog(ctx context.Context, req *workflowexecutionv1.GetEventLogRequest) (*workflowexecutionv1.GetEventLogResponse, error) {
	if req == nil || req.ExecutionId == "" {
		return nil, grpclib.InvalidArgumentError("execution_id is required")
	}

	pageSize := int(req.PageSize)
	if pageSize <= 0 {
		pageSize = defaultEventPageSize
	}
	if pageSize > maxEventPageSize {
		pageSize = maxEventPageSize
	}

	// Map event_types filter to a single string for the store query.
	// The store supports filtering by one event type; if the request has
	// multiple types, we fetch all and filter in memory.
	var eventTypeFilter string
	if len(req.EventTypes) == 1 {
		eventTypeFilter = req.EventTypes[0].String()
	}

	// Fetch one extra record to determine has_more
	records, err := c.store.GetWorkflowExecutionEvents(
		ctx,
		req.ExecutionId,
		int64(req.AfterSequence),
		eventTypeFilter,
		req.TaskName,
		pageSize+1,
	)
	if err != nil {
		log.Error().Err(err).
			Str("execution_id", req.ExecutionId).
			Msg("Failed to query execution events")
		return nil, grpclib.InternalError(err, "failed to query execution events")
	}

	hasMore := len(records) > pageSize
	if hasMore {
		records = records[:pageSize]
	}

	// Build multi-type filter set for in-memory filtering when >1 event type is requested
	var typeFilter map[string]struct{}
	if len(req.EventTypes) > 1 {
		typeFilter = make(map[string]struct{}, len(req.EventTypes))
		for _, et := range req.EventTypes {
			typeFilter[et.String()] = struct{}{}
		}
	}

	events := make([]*workflowexecutionv1.WorkflowExecutionEvent, 0, len(records))
	var latestSeq uint64

	for _, rec := range records {
		// Apply multi-type filter if needed
		if typeFilter != nil {
			if _, ok := typeFilter[rec.EventType]; !ok {
				continue
			}
		}

		evt := &workflowexecutionv1.WorkflowExecutionEvent{}
		if err := proto.Unmarshal(rec.Data, evt); err != nil {
			log.Warn().Err(err).
				Str("execution_id", req.ExecutionId).
				Int64("sequence_number", rec.SequenceNumber).
				Msg("Skipping malformed event record")
			continue
		}

		events = append(events, evt)
		if uint64(rec.SequenceNumber) > latestSeq {
			latestSeq = uint64(rec.SequenceNumber)
		}
	}

	return &workflowexecutionv1.GetEventLogResponse{
		Events:         events,
		HasMore:        hasMore,
		LatestSequence: latestSeq,
	}, nil
}

File: backend/services/stigmer-server/src/domain/workflowexecution/get-event-log.ts
/**
 * GetEventLog — ports get_event_log.go: cursor-paginated reads over the
 * workflow_execution_events side table (sequence_number ascending,
 * strictly after the cursor). The asymmetry the Class A suite pins: an
 * empty execution_id refuses InvalidArgument, but an UNKNOWN id answers
 * an empty page — there is deliberately no existence check (the opposite
 * of the subscribe lanes).
 *
 * The store filters by at most one event type; multi-type requests fetch
 * unfiltered and filter in memory — which can under-fill a page (Go
 * accepts this: has_more is computed from the pre-filter fetch).
 */
import { create, fromBinary } from "@bufbuild/protobuf";

import { WorkflowExecutionEventSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/event_pb";
import type { WorkflowExecutionEvent } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/event_pb";
import { WorkflowEventType } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/event_pb";
import { GetEventLogResponseSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";
import type {
  GetEventLogRequest,
  GetEventLogResponse,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";

import type { Logger } from "../../boot/logger.js";
import { internalError, invalidArgumentError } from "../../pipeline/errors.js";
import type { Store, WorkflowExecutionEventRecord } from "../../store/interface.js";

import { DEFAULT_EVENT_PAGE_SIZE, MAX_EVENT_PAGE_SIZE } from "./constants.js";

export interface EventLogDeps {
  readonly store: Store;
  readonly logger: Logger;
}

/** The proto enum's value NAME — exactly Go's WorkflowEventType.String(). */
export function eventTypeName(eventType: WorkflowEventType): string {
  return WorkflowEventType[eventType] ?? "";
}

export async function getEventLog(
  deps: EventLogDeps,
  req: GetEventLogRequest,
): Promise<GetEventLogResponse> {
  if (req.executionId === "") {
    throw invalidArgumentError("execution_id is required");
  }

  let pageSize = req.pageSize;
  if (pageSize <= 0) {
    pageSize = DEFAULT_EVENT_PAGE_SIZE;
  }
  if (pageSize > MAX_EVENT_PAGE_SIZE) {
    pageSize = MAX_EVENT_PAGE_SIZE;
  }

  // The store filters by one event type; multiple requested types fall
  // back to in-memory filtering after an unfiltered fetch.
  let eventTypeFilter = "";
  if (req.eventTypes.length === 1) {
    eventTypeFilter = eventTypeName(req.eventTypes[0]);
  }

  // Fetch one extra record to compute has_more.
  let records: WorkflowExecutionEventRecord[];
  try {
    records = await deps.store.getWorkflowExecutionEvents(
      req.executionId,
      Number(req.afterSequence),
      eventTypeFilter,
      req.taskName,
      pageSize + 1,
    );
  } catch (error) {
    deps.logger.error("Failed to query execution events", {
      executionId: req.executionId,
      error: error instanceof Error ? error.message : String(error),
    });
    throw internalError(error, "failed to query execution events");
  }

  const hasMore = records.length > pageSize;
  if (hasMore) {
    records = records.slice(0, pageSize);
  }

  let typeFilter: Set<string> | undefined;
  if (req.eventTypes.length > 1) {
    typeFilter = new Set(req.eventTypes.map(eventTypeName));
  }

  const events: WorkflowExecutionEvent[] = [];
  let latestSequence = 0n;

  for (const record of records) {
    if (typeFilter !== undefined && !typeFilter.has(record.eventType)) {
      continue;
    }

    let event: WorkflowExecutionEvent;
    try {
      event = fromBinary(WorkflowExecutionEventSchema, record.data);
    } catch {
      deps.logger.warn("Skipping malformed event record", {
        executionId: req.executionId,
        sequenceNumber: record.sequenceNumber,
      });
      continue;
    }

    events.push(event);
    if (BigInt(record.sequenceNumber) > latestSequence) {
      latestSequence = BigInt(record.sequenceNumber);
    }
  }

  return create(GetEventLogResponseSchema, {
    events,
    hasMore,
    latestSequence,
  });
}
